core/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -194,23 +193,24 @@ protected boolean isBarrierNode(ReaderNode node) {
 
     /**
      * This method is called during the second pass of {@link WaySegmentParser} and provides an entry point to enrich
-     * the given OSM way with additional tags before it is split into segments.
+     * the given OSM way with additional tags before it is passed on to the tag parsers.
      */
-    protected void setArtificialWayTags(GHPoint first, GHPoint last, ReaderWay way) {
-        // todo: should we not rather do all this stuff **per edge** not per way? might be a bit slower, but we
-        // could also do raster optimization in country lookup if this was the bottleneck.
-
+    protected void setArtificialWayTags(PointList pointList, ReaderWay way) {
         // Estimate length of ways containing a route tag e.g. for ferry speed calculation
-        double firstLat = first.getLat(), firstLon = first.getLon();
-        double lastLat = last.getLat(), lastLon = last.getLon();
+        double firstLat = pointList.getLat(0), firstLon = pointList.getLon(0);
```

Comment:
Sadly, because we calculate this estimated center we may choose another country (see https://discuss.graphhopper.com/t/country-of-way-is-wrong-on-road-near-border-with-curvature/6908/2). Wouldn't be a better idea to pick up a middle point (or close to is) as:

```java
double middleLat = pointList.getLat(pointList.size() / 2), middleLon = pointList.getLon(pointList.size() / 2);
```

This way we make sure the selected point in on the edge/way.

Reply:
I opened https://github.com/graphhopper/graphhopper/issues/2467 for this